docs/ECO-METADATA.md

```diff
@@ -0,0 +1,46 @@
+# EVE meta-data server for app instances
+
+EVE provides an old method to serve cloud-init meta-data in the form of a read-only CDROM disk which is created on the fly when an app instance is booted if the API specifies userData/cipherData in the [AppInstanceConfig message](../api/proto/config/appconfig.proto).
+
+However, there is also a need to provide access to meta-data which might change while the app instance is running.
+
+The first data which is needed by applications is determining the external IP address of the edge node, so that when there is a portmap ACL in place the application instance can determine which IP plus port its peers should use to connect to it.
+
+The initial meta-data service provides merely that, but over time we expect to add the rest of the cloud-init content.
+
+## Schema
+
+There is no existing industry standard schema specifying a notion of an external IP; existing schemas contain public and private IP addresses but the external IP is a different thing necessitated by the internal NAT which EVE deploys for the local network instances.
+
+Thus this particular part of the meta-data uses a EVE-unique schema, which we do not expect for other meta-data information.
+
+The API endpoint is <http://169.254.169.254/eve/v1/network.json>
```

Comment:
So yes -- this particular URL pretty much flatly precludes us from trying to sneakily enable cloud-init support -- we'll have to write our own cloud init-provider (and good luck updating all the existing VM images with that!).

It is not the end of the world -- and no reason not to merge this particular PR, but I'm thinking that we should REALLY be thinking hard of picking one datasource from this list https://cloudinit.readthedocs.io/en/latest/topics/datasources.html and spoofing it to get cloud-init support for free.

In fact, now that you've done the hard part (network plumbing) I can ask @giggsoff &co to actually implement a spoof URL.

All of this to say if the above strategy works -- we may completely obviate the need for http://169.254.169.254/eve

Reply:
@rvs we should definitely use (one of the many) standard URLs for the cloud-init meta-data and user-data, whether that is the AWS http://169.254.169.254/2009-04-04/meta-data/ or some other.

But that does not obviate the need for external_ipv4, since none of the cloud init schemas I've looked at has a notion of an external IP. They have a notion of a local IP (which is the IP address assigned to the workload) and a public IP (which is an IP on the public Internet outside of some external NAT), but not this middle thing which EVE brings due to the implicit NAT in the local network instance with its port map. I think it would be a big mistake to usurp the "public IP" in an existing schema to mean the "external IP" in EVE.